src/helpers/sso-helper.js

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+ * See LICENSE in the project root for license information.
+ */
+
+import { dialogFallback } from "./fallbackauthdialog.js";
+import { getUserData } from "./middle-tier-calls";
+import { showMessage } from "./message-helper";
+import { handleClientSideErrors } from "./error-handler";
+
+/* global OfficeRuntime */
+
+let retryGetMiddletierToken = 0;
+
+export async function getUserProfile(callback) {
+  try {
+    let middletierToken = await OfficeRuntime.auth.getAccessToken({
+      allowSignInPrompt: true,
+      allowConsentPrompt: true,
+      forMSGraphAccess: true,
+    });
+    let response = await getUserData(middletierToken);
+    if (!response) {
+      throw new Error("Middle tier didn't respond");
+    } else if (response.claims) {
+      // Microsoft Graph requires an additional form of authentication. Have the Office host
+      // get a new token using the Claims string, which tells AAD to prompt the user for all
+      // required forms of authentication.
+      let mfaMiddletierToken = await OfficeRuntime.auth.getAccessToken({
+        authChallenge: response.claims,
+      });
+      response = getUserData(mfaMiddletierToken);
+    }
+
+    // AAD errors are returned to the client with HTTP code 200, so they do not trigger
+    // the catch block below.
+    if (response.error) {
+      handleAADErrors(response, callback);
+    } else {
+      callback(response);
+    }
+  } catch (exception) {
+    // if handleClientSideErrors returns true then we will try to authenticate via the fallback
+    // dialog rather than simply throw and error
+    if (exception.code) {
+      if (handleClientSideErrors(exception)) {
+        dialogFallback(callback);
+      }
+    } else {
+      showMessage("EXCEPTION: " + JSON.stringify(exception));
+      throw exception;
+    }
+  }
+}
+
+function handleAADErrors(response, callback) {
+  // On rare occasions the middle tier token is unexpired when Office validates it,
```

Comment:
Not super important, but to me It's a little confusing to call it the "middle tier token". I just call it an access token. Or access token to middle tier.

Reply:
I wanted to distinguish between the token to access to the middle tier and the token to access graph (in the middle tier).  I was confused by this when trying to make the changes so I wanted to include a name that would help understand what is happening.